### PR TITLE
fix: pad non-square resize outputs to multiples of patch_size * num_windows

### LINKED ISSUE #983: When square_resize_div_64 = False is set, an error occurs.
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

一、Problem
When processing downstream tasks, I want to try not changing the aspect ratio of the input image, so I set square_resize_div_64 = False.
However, the following error occurs.

Image

While examining the implementation of the function _build_train_resize_config in coco.py, I found that this parameter ultimately affects the parameter shown in the figure: "LongestMaxSize".

Image

After modifying this parameter to 2000, the error changed to:

Image

I set multi_scale = False, so the max size of the shorter side should be 896, but the actual error shows that the shorter side exceeds 896.

### Environment

RF-DETR: Latest

### Minimal Reproducible Example

 model = RFDETRMedium(
        num_windows=2,      
        patch_size=16,     
        resolution=896,     
        num_classes=3,
        positional_encoding_size = 896//16,
    )

    model.train(
        # dataset_file = 'coco',
        #dataset_dir = r'D:\aotto\quexian-jiance\data-20260331\coco', 
         multi_scale = False,
        square_resize_div_64 = False
    )

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes, I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Irfan-Hamid**
## What does this PR do?

Fixes an error when training with `square_resize_div_64=False`. The non-square training and validation resize pipelines (`SmallestMaxSize` then `LongestMaxSize`) did not guarantee output dimensions divisible by `patch_size * num_windows`, causing `WindowedDinov2WithRegistersEmbeddings.forward` to reject inputs with `"Input spatial dimensions must be divisible by patch_size * num_windows"`. Appends `PadIfNeeded` (with `pad_*_divisor` set to `patch_size * num_windows`) after the resize pair in both the train (`_build_train_resize_config`) and val/test branches of `make_coco_transforms`.

**Why padding is safe for the backbone:**

- `interpolate_pos_encoding` in DINOv2 is already dynamic, so it handles any `H x W` divisible by `patch_size`.
- The backbone encoder does not receive the `NestedTensor` mask (only downstream attention does), and the batch collate step already zero-pads images up to batch-max size. The divisor-pad is just one more small strip (at most `divisor - 1` pixels) on top of that existing padding.
- No changes to backbone, `NestedTensor`, windowed attention, or detection head.

**Fix and pipeline wiring:**

- Added private helper `_pad_to_divisor_config(divisor)` that builds an Albumentations `PadIfNeeded` config, with `min_height`/`min_width` set to `None` to avoid the 1024 default colliding with `pad_*_divisor`.
- Added `divisor: Optional[int] = None` kwarg to `_build_train_resize_config`. When set and `square=False`, the helper appends `PadIfNeeded` to both OneOf options.
- `make_coco_transforms` computes `block_size = patch_size * num_windows` and threads it into the train and val/test pipelines.
- The square branch (`make_coco_transforms_square_div_64`) and the `val_speed` branch are untouched since they already produce compliant dimensions.

**Related Issue(s):** Fixes #983

## Type of Change

- Bug fix (non-breaking change that fixes an issue)

## Testing

- [x] I have tested this change locally
- [x] I have added/updated tests for this change

**Test details:**

- Added `TestBuildTrainResizeConfigDivisor` (5 structural tests) covering the new `divisor` kwarg behavior for both non-square and square branches, plus backward compatibility when `divisor=None`.
- Added `TestNonSquareResizeDivisibilityRegression` (8 end-to-end regression tests) that feed images of varied aspect ratios (portrait, landscape, wide, tall, arbitrary non-aligned) through the resize wrapper and through `make_coco_transforms` (train and val) across 40 random seeds, asserting output `H` and `W` are divisible by 32 for `patch_size=16, num_windows=2`.
- Updated two existing count assertions in `tests/datasets/test_augmentations.py` since the val/test pipeline now has 3 resize wrappers instead of 2.
- All 271 tests in `tests/datasets/` pass locally.
- `pre-commit run` clean on touched files.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [x] I have updated the documentation accordingly (if applicable)

## Additional Context

The `max_size=1333` hardcoded cap in the non-square resize path is a separate latent concern (for very elongated aspect ratios it can shrink the shorter side below the configured resolution) but is out of scope here. This PR is narrowly focused on the divisibility bug in #983.

### CODE REVIEW

**Comment by Irfan-Hamid** [comment]
Closing for now. Cannot sign the CLA currently.
